### PR TITLE
Allow different floating point types in hierarchy

### PR DESCRIPTION
**Pull request by aprokop**
#916 made me realize that we never tested whether we are able to actually run a hierarchy built using `double`. Turns out we didn't. This patch fixes it, and updates one of the benchmarks (brute vs bvh) to run both float and double.

Scope:

- Fixes few places in the hierarchy where we unconditionally used `Box<DIM>` and similar
- Updates a benchmark

### CODE REVIEW

**Review by masterleinad — COMMENTED** [review]
Looks like this is based on top of #921 now. Otherwise, this looks good to me.

**Comment by aprokop** [comment]
> Looks like this is based on top of #921 now. Otherwise, this looks good to me.

Yeah, I'll rebase it. Decided to move out the coordinate things first.

**Review by masterleinad — COMMENTED** [review]
Is there any particular reason that you chose the brute force benchmark?

**Comment by aprokop** [comment]
It uses both BVH and BruteForce, so i could check that it works for both of them.

**Review by masterleinad — APPROVED** [review]
Fine with me.

File: src/details/ArborX_DetailsMortonCode.hpp [hunk 1]
```diff
@@ -316,7 +316,7 @@ KOKKOS_INLINE_FUNCTION unsigned long long morton64(Point const &p)
   unsigned long long r = 0;
   for (int d = 0; d < DIM; ++d)
   {
-    auto x = min(max(p[d] * N, 0.f), (float)N - 1);
+    auto x = min(max((float)p[d] * N, 0.f), (float)N - 1);
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsMortonCode.hpp:319]: I think this is fine. This is the case of 64-bit and dimension >= 3. The number of bits used for this Morton codes is <= 21.
- We don't use any additions of floating point numbers for calculation, which means that the machine precision is not in play.
-  The smallest integer not represented by `float` is $2^24$, which means that all the integers in the range $[0, N)$ are represented by floats.

File: src/ArborX_LinearBVH.hpp [hunk 2]
```diff
@@ -178,6 +181,8 @@ BasicBoundingVolumeHierarchy<MemorySpace, Value, IndexableGetter,
                                               ExecutionSpace>::value,
                 "Primitives must be accessible from the execution space");
   constexpr int DIM = GeometryTraits::dimension_v<BoundingVolume>;
+  using FloatingPoint =
+      typename GeometryTraits::coordinate_type<BoundingVolume>::type;
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:185]: I would prefer if you deduced these closer to the place you actually use them.  As far as I can tell there is really one place.

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
Gave up on templating `Placeholder` on `Point`, with different compilers either doing ok or failing. Not going to waste more time on this.

**Pull request merged**